odoo/http.py

```diff
@@ -945,6 +956,132 @@ def touch(self):
         self.is_dirty = True
 
 
+
+# =========================================================
+# GeoIP
+# =========================================================
+
+class GeoIP(collections.abc.Mapping):
+    """
+    Ip Geolocalization utility, determine information such as the
+    country or the timezone of the user based on his IP Address.
+
+    The instances share the same API as `:class:`geoip2.models.City`
+    <https://geoip2.readthedocs.io/en/latest/#geoip2.models.City>`_.
+
+    When the IP couldn't be geolocalized (missing database, bad address)
+    then an empty object is returned. This empty object can be used like
+    a regular one with the exception that all info are set None.
+
+    :param str ip: The IP Address to geo-localize
+
+    .. note:
+
+        The geoip info the the current request are available at
+        :attr:`~odoo.http.request.geoip`.
+
+    .. code-block:
+
+        >>> GeoIP('127.0.0.1').country.iso_code
+        >>> GeoIP('178.33.40.43').country.iso_code  # odoo.com
+        'FR'
+    """
+
+    # Two empty objects used when the geolocalization failed. They have
+    # the sames attributes as real countries/cities except that
+    # accessing them evaluates to None.
+    empty_country = geoip2.models.Country({})
+    empty_city = geoip2.models.City({})
```

Comment:
why class attributes?
constants wouldn't work? they're immutable, right?

Reply:
Constant would work indeed, I just like it is namespaced with GeoIP. RCO did the same comment, apparently it confuses people so I'll move it to the global scope